File: gfPlayer.java
import java.util.*;

class gfPlayer {
  //Fields
  ArrayList<GFCard> hand;

  private int handValue;
  private int bookCount = 0;
  String playerName;
  boolean player;
  boolean turn;

  String choose;// for the player
  int choose1;  // for the card
  Scanner scan = new Scanner (System.in);

  //Constructor
  gfPlayer(boolean p, String n){
    hand = new ArrayList<GFCard>();

    handValue = 0;
    player = p;
    playerName = n;
  }

  //Methods

  //Kinda new
  public void setStartHand(GFCard card){ // This adds a card to the hand at the start
    hand.add(card);
    //GoFish.incrementTopCard();
  }

  //Not new
  public void printHand(){
    if(player){ // prints your hand and how many cards you have
      sortHand();
      System.out.println("You have: " + getHandSize() + " cards, " + bookCount + " books");
      for(GFCard c : hand){
        c.printGFCard();
        System.out.print (", ");
      }
      System.out.println();
    } else if (!player){ // it prints the amount of cards in the hand for "AI" players
      System.out.println(playerName + " has: " + getHandSize() + " cards, " + bookCount + " books" );
    }
  }


  //This is new
  public String chooseWho(){
    if (player){ // This asks the player who they want to ask for a card
      System.out.println ("  Who would you like to ask for a card");
      choose = scan.nextLine().toLowerCase();
      checkChooseWho();
      return choose;
    }else{
      choose = "player " + (int) (Math.random()*4 + 1);// the AI just ask a random person
      checkChooseWho();
      if(choose.equals("player 4")) choose = "you";
      System.out.print("  They asked " + choose + " for a ");
      return choose;
    }
  }

  //This is new
  public int chooseCard(){
    if (player){ // This asks the player what card he wants to ask for
      Scanner sc = new Scanner (System.in); // face cards have to be thier value. I don't know how else to do it
      System.out.println ("  What card would you like to ask for?");
      choose1 = sc.nextInt();
      return choose1;
    }else{
      choose1 =  (int) (Math.random()*13 + 1) ; // the AI players just ask for a random card
      System.out.print (choose1);
      System.out.println();
      return choose1;
    }
  }

  //This is new
  public boolean hasCard(int card){ // checks to see if the player has the card they want
    for (int i = 0; i<getHandSize(); i++){
      if (hand.get(i).getValue() == card){
        return true;
      }
    }
    return false;
  }

  //This is new
  void checkChooseWho(){
    int i = Integer.parseInt(choose.substring(7));
    if (player){
      if (choose.substring(0,7).equals ("player ")){}
      else {
        System.out.println ("Did you even read the Instructions?");
        choose = scan.nextLine().toLowerCase();
      }
      if (i >= 4){
        System.out.println ("Thats not a player stupid");
        choose = scan.nextLine().toLowerCase();
      }
    }else {
      if (choose.equals(playerName)){
        choose = "player " + (int) (Math.random()*4 + 1);// the AI just ask a random person
      }
    }

  }

  //This is new
  public void checkForBooks(){ // checks to see if you have 2 of the same cards and then removes them
    sortHand();                // A pair is called a book in Go Fish for some unknow reason
    int t = bookCount;
    for (int i = 0; i < getHandSize(); i++){ // not sure if this works or not
      for (int j = 0; j < getHandSize(); j++){
        if (i != j &&getHand().get(i).getValue() == getHand().get(j).getValue()) {
          hand.remove(j);
          hand.remove(i);
          i = 0;
          j = 0;
          bookCount++;
          //if (player && t != bookCount) System.out.println("You got another book");
        }
      }
    }
  }

  //Not my code
  void sortHand(){ //No idea how this works but it does
    Collections.sort (hand, new Comparator<GFCard>(){
      @Override
      public int compare(GFCard card1, GFCard card2){
        Integer x1 = card1.getValue();
        Integer x2 = card2.getValue();
        return  x1.compareTo( x2);
      }
    }
    );
  }

  //Not new
  void goFish(GFCard c){ //This is the same as setStartHand becasue i'm too lazy to go and change it
    hand.add(c); // adds a card to your hand
  }

  //Accessor method
  int getHandSize (){ // gets the number of cards in hand
    return hand.size();
  }

  //Accessor method
  int getBookCount(){
    return bookCount;
  }

  //Accessor method
  public ArrayList<GFCard> getHand(){ // gets the cards in your hand
    return hand;
  }
}

File: gfPlayer.js
function GFPlayer(player, playerName) {
  return {
    hand: [],
    handValue: 0,
    p: player,
    n: playerName,
    choose1: 0, //card
    choose: '', //player
    card: 0,
    bookCount: 0,

    setStartHand: function (card) {
      this.hand.push(card);
      if (this.p)card.printGFCard();
    }, //setStartHand

    printHand: function () { // prints your hand and how many cards you have
      //sort the hand using sortHand();
      // if player show hand size not as a button and cards as buttons
      if (this.p) {
        for (var i = 0; i < this.hand.length; i++) {
          var c = this.hand[i];
          console.log(c);
          this.printHandSize();
          this.printPairArmount();
          c.printGFCard();
        } //for
      } else if (!this.p) { //if not the player
        //TODO: why is this in 2 seperate ifs. have hand size and pair amount no matter what
        this.printHandSize();
        this.printPairArmount();
      } //else if (!this.p)
    }, //printHand

    printHandSize: function () {
      //prints the amount of cards in each players hand
      //definitly could be more efficient
      var x;
      if (this.p) { //player 0
        x = document.getElementById('player-hand-size');
        x.innerHTML = this.hand.length + ' cards ';
        document.getElementById('player-info').appendChild(x);
      }else if (this.n == 'player 1') { //player 1
        x = document.getElementById('player1-hand-size');
        x.innerHTML = this.hand.length + ' cards ';
        document.getElementById('player1-info').appendChild(x);
      }else if (this.n == 'player 2') { //player 2
        x = document.getElementById('player2-hand-size');
        x.innerHTML = this.hand.length + ' cards ';
        document.getElementById('player2-info').appendChild(x);
      }else { //player 3
        x = document.getElementById('player3-hand-size');
        x.innerHTML = this.hand.length + ' cards ';
        document.getElementById('player3-info').appendChild(x);
      } //else
    }, //printHandSize

    printPairArmount: function () { // prints the amount of pairs each player has
      //TODO: add an id to each player for design? back of the card for how many pairs you have?
      var x;
      if (this.p) { //player 0
        x = document.getElementById('player-pair-count');
        x.innerHTML = this.bookCount + ' pairs';
        document.getElementById('player-info').appendChild(x);
      }else if (this.n == 'player 1') { //player 1
        x = document.getElementById('player1-pair-count');
        x.innerHTML =  this.bookCount + ' pairs';
        document.getElementById('player1-info').appendChild(x);
      } else if (this.n == 'player 2') { //player 2
        x = document.getElementById('player2-pair-count');
        x.innerHTML =  this.bookCount + ' pairs';
        document.getElementById('player2-info').appendChild(x);
      }else if (this.n == 'player 3') { //player 3. why is this an if else and not just an else
        x = document.getElementById('player3-pair-count');
        x.innerHTML = this.bookCount + ' pairs';
        document.getElementById('player3-info').appendChild(x);
      }
    },

    chooseWho: function () {
        choose = 'player ' + Math.floor(Math.random() * 4);
        this.checkChooseWho(); //makes sure the player doesn't ask himself for a  card
        return choose;
      },

    checkChooseWho: function () {
        if (choose == this.n) {  //if(this.choose.equals(playerName))
          this.chooseWho();//asks a random person
        } //if
      }, //checkChooseWho

    chooseCard: function () { // This is going to have to change to button inputs
      choose1 = Math.floor(Math.random() * 13 + 1);
      return choose1;
    }, //chooseCard

    checkChooseCard: function () { //checks to see if the input is an actual card
      if (!isNaN(sc)) choose1 = sc; //if the sc scanner isn't a number run again
      else {
        console.log("  Opps! That isn't a card. Try again.");
        chooseCard();
        //If the input isn't a number it reruns
      } //else

      if (choose1 > 13) { //reruns if greater than 13
        console.log("  Opps! That isn't a card. Try again.");
        chooseCard();
      } //if choose1 > 13
    }, //checkChooseCard

    hasCard: function () { //checks to see if there is a match. returns true of there is
      for (var i = 0; i < this.hand.length; i++) {
        if (this.hand[i].getValue() == card) {
          return true;
        } //if
      } //for i

      return false;
    }, //hasCard

    checkForBooks: function () {
      var l; //l for log
      for (var i = 0; i < this.hand.length; i++) {
        for (var j = 0; j < this.hand.length; j++) {
          //TODO: change this to a while loop in case there is a pair with the first number
          // console.log(j + '' + i);
          // console.log(this.getHand()[i].getValue());
          // console.log(this.getHand()[j].getValue());
          if (i != j && this.getHand()[i].getValue() == this.getHand()[j].getValue()) {
            //// TODO: set variables for this.getHand()[i].getValue() and such?
            //removes the card buttons \/
            if (this.p) {
              document.getElementById(
                this.getHand()[j].getValue() + this.getHand()[j].getSuit()
              ).remove();
              document.getElementById(
                this.getHand()[i].getValue() + this.getHand()[i].getSuit()
              ).remove();

              //logs that you got a pair \/
              l = document.createElement('p');
              l.innerHTML = 'You got a pair of ' + this.getHand()[i].getValue() + 's';
              l.scrollTop = l.scrollHeight;
              document.getElementById('log').appendChild(l);
            } //if player

            this.hand.splice(j, 1);  //removes the pair
            this.hand.splice(i, 1);  //removes the pair
            i = 0; //sets it back in case there are more than 1 pair
            j = -1; //-1 because it will increase back to 0 before it loops again. makes sure
            this.bookCount++;
          } //if values == each other
        } //for j
      } //for i
    }, //checkForBooks

    sortHand: function () {
      //this.hand.sort();

      //TODO: find a way to sort the hand
      /* this is the old code
      Collections.sort (hand, new Comparator<GFCard>(){
        @Override
        public int compare(GFCard card1, GFCard card2){
          Integer x1 = card1.getValue();
          Integer x2 = card2.getValue();
          return  x1.compareTo( x2);
          }
        }
      );
      */
    }, //sortHand

    removeCard: function () { //// TODO: Useless? yes
      if (this.p)  document.getElementById(this.card).remove();
    },//removeCard

    goFish: function (card) { //adds the card to the players hand then displays it
      this.hand.push(card);
      if (this.p) card.printGFCard();
    }, //goFish

    getHandSize: function () { //Useless? same as printHandSize?
      var x = document.getElementById('player-hand-size');
      x.setAttribute('value', 'hi');
      document.body.appendChild(x);
    }, //getHandSize

    getBookCount: function () {
      return this.bookCount;
    }, //getBookCount

    getHand: function () {
      return this.hand;
    }, //getHand

  }; //return
} //function
